### PR TITLE
Support Picolibc in addition to newlib

### PR DESCRIPTION
**Pull request by keith-packard**
## Add support for picolibc

Picolibc is a C library that is designed to work on 32 and 64 bit embedded systems. It is available on Debian derivatives, can be built using crosstool-ng or can be added to the GNU Arm Embedded Toolkit.

Picolibc is comprehensively tested on Arm, Aarch64, x86 and RISC-V embedded systems to ensure standards conformance. It offers significant memory savings in typical usage as well as avoiding dynamic memory allocation for even complex code paths like floating point conversions.

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in the submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use a web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [x] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [ ] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [ ] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [ ] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality
- [x] 本拉取/合并使用[formatting](https://github.com/mysterywolf/formatting)等源码格式化工具确保格式符合[RT-Thread代码规范](../documentation/contribution_guide/coding_style_cn.md) This PR complies with [RT-Thread code specification](../documentation/contribution_guide/coding_style_en.txt)

### CODE REVIEW

**Comment by keith-packard** [comment]
> [CLA assistant check](https://cla-assistant.io/RT-Thread/rt-thread?pullRequest=5762) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RT-Thread/rt-thread?pullRequest=5762) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RT-Thread/rt-thread?pullRequest=5762) it.

I cannot get permission from my employer for any CLA.

**Comment by mysterywolf** [comment]
Thank you for your contribution!

File: include/rtdef.h [hunk 1]
```diff
@@ -606,6 +606,14 @@ struct rt_cpu
 
 #endif
 
+#if defined(RT_USING_PICOLIBC) && defined(_HAVE_PICOLIBC_TLS_API)
+#define RT_USING_TLS
```
Comment (BernardXiong) [inline include/rtdef.h:610]: So good for this feature. It's a feature planning in v5.0.

File: src/thread.c [hunk 2]
```diff
@@ -177,12 +180,22 @@ static rt_err_t _thread_init(struct rt_thread *thread,
     /* init thread stack */
     rt_memset(thread->stack_addr, '#', thread->stack_size);
 #ifdef ARCH_CPU_STACK_GROWS_UPWARD
+#ifdef RT_USING_TLS
+    thread->tls = thread->stack_addr;
+    _init_tls(thread->tls);
```
Comment (BernardXiong) [inline src/thread.c:185]: From my understanding, it's a porting layer for tls (in libc?). It's better to use a unify name for here.

**Review by BernardXiong — COMMENTED** [review]
Thank you for your contribution. But, where is picolibc implementation? Is it made as a software package?

**Comment by keith-packard** [comment]
> Thank you for your contribution. But, where is picolibc implementation? Is it made as a software package?

Oh, Sorry! Yes, picolibc is available at https://github.com/picolibc/picolibc -- the [current version is 1.7.6](https://github.com/picolibc/picolibc/releases/tag/1.7.6). I've given numerous presentations:

https://www.youtube.com/watch?v=SC6aBezNFFQ
https://www.youtube.com/watch?v=qhavUyWJ46s
https://www.youtube.com/watch?v=K9C4AJZKauE

Picolibc is also available in Debian and Debian derivatives, and is included in [Crosstool-ng](https://github.com/crosstool-ng/crosstool-ng). Zephyr has recently approved including Picolibc as a module, which means it can be built along with the operating system.

File: src/thread.c (hunk 2 above)
Comment (keith-packard) [reply src/thread.c:185]: Agreed -- I will adjust this to use an rt-prefixed name instead of the internal picolibc name.

Comment (keith-packard) [reply src/thread.c]: I've pushed an updated version which creates an rt abstraction and uses the picolibc implementation for picolibc.

**Comment by keith-packard** [comment]
Oh, I should have made it more clear that the three patches modifying existing BSP are meant as demonstrations of how to use picolibc and what the benefits might be. I would not expect those to be merged unless RT-Thread is ready to accept picolibc as a hard dependency.

File: src/thread.c (hunk 2 above)
Comment (keith-packard) [reply src/thread.c]: And another update which installs the GNU Arm Embedded Toolkit version of picolibc for CI tests.

**Comment by keith-packard** [comment]
Not sure what happened, but I got the wrong URL for the picolibc bits somehow (an extra 'v' snuck in). Weird.

**Comment by mysterywolf** [comment]
Sorry, We cannot merged it PR. Thanks for you contribution, again.

**Pull request closed without merging**